File: openapi.jl
import YAML: YAMLDocIterator, iterate, load_all_file
import URIs: uristring

mutable struct Tag <: Comparable
    name::String
    description::String
    externalDocs::ExternalDocumentation
end

Tag() = Tag("", "", ExternalDocumentation())

function parse!(t::Tag, data::Dict{Any,Any})
    for (key,value) in data
        if key == "name" t.name = value end
        if key == "description" t.description = convert(value) end
        if key == "externalDocs" parse!(t.externalDocs, value) end
    end
end

function parse!(v::Vector{Tag}, data::Vector{Dict{Any,Any}})
    for d in data
        t = Tag()
        parse!(t, d)
        push!(v, t)
    end
end

mutable struct Components <: Comparable
    schemas::Dict{String,Schema}
    responses::Dict{String,Response}
    parameters::Dict{String,Parameter}
    examples::Dict{String,Example}
    requestBodies::Dict{String,RequestBody}
    headers::Dict{String,Header}
    securitySchemes::Dict{String,SecurityScheme}
end

Components() = Components(Dict{String,Schema}(), Dict{String,Response}(), Dict{String,Parameter}(),
    Dict{String,Example}(), Dict{String,RequestBody}(), Dict{String,Header}(), Dict{String,SecurityScheme}())

function parse!(c::Components, data::Dict{Any,Any})
    for (key,value) in data
        if key == "schemas"
            for (k,v) in value
                s = Schema()
                parse!(s, v)
                c.schemas[k] = s
            end
        end
        if key == "responses"
            for (k,v) in value
                r = Response()
                parse!(r, v)
                c.responses[k] = r
            end
        end
        if key == "parameters"
             for (k,v) in value
                 p = Parameter()
                 parse!(p, v)
                 c.parameters[k] = p
             end
         end
        if key == "example"
            for (k,v) in value
                e = Example()
                parse!(e, v)
                c.examples[k] = e
            end
        end
        if key == "requestBodies"
            for (k,v) in value
                r = RequestBody()
                parse!(r, v)
                c.requestBodies[k] = r
            end
        end
        if key == "headers"
            for (k,v) in value
                h = Header()
                parse!(h, v)
                c.headers[k] = h
            end
        end
        if key == "securitySchemes"
            for (k,v) in value
                s = SecurityScheme()
                parse!(s, v)
                c.securitySchemes[k] = s
            end
        end
    end
end

mutable struct TagGroup <: Comparable
    name::String
    tags::Vector{String}
end

TagGroup() = TagGroup("", Vector{String}())

function parse!(t::TagGroup, data::Dict{Any,Any})
    for (key,value) in data
        if key == "name" t.name = value end
        if key == "tags"
            for tg in value push!(t.tags, tg) end
        end
    end
end

function parse!(t::Vector{TagGroup}, data::Vector{Dict{Any,Any}})
    for d in data
        tg = TagGroup()
        parse!(tg, d)
        push!(t, tg)
    end
end

mutable struct OpenAPI <: Comparable
    openapi::String
    info::Info
    jsonSchemaDialect::String
    servers::Vector{Server}
    paths::Dict{String,PathItem}
    webhooks::Dict{String,PathItem}
    components::Components
    security::Vector{SecurityRequirement}
    tags::Vector{Tag}
    externalDocs::ExternalDocumentation
    tagGroups::Vector{TagGroup}
end

OpenAPI() = OpenAPI("", Info(), "", Vector{Server}(), Dict{String,PathItem}(), Dict{String,PathItem}(),
    Components(), Vector{SecurityRequirement}(), Vector{Tag}(), ExternalDocumentation(), Vector{TagGroup}())

function parse(it::YAMLDocIterator)::OpenAPI
    (yaml, state) = iterate(it)
    @debug "YAML iterator state: $state"

    api = OpenAPI()

    for (key, value) in yaml
        @info "key: $(key); key-type: $(typeof(key)); value-type: $(typeof(value))"
        if key == "openapi" api.openapi = value end
        if key == "info" parse!(api.info, value) end
        if key == "jsonSchemaDialect" api.jsonSchemaDialect = value end
        if key == "servers" parse!(api.servers, value) end
        if key == "paths"
            for (k,v) in value
                pi = PathItem()
                parse!(pi, v)
                api.paths[k] = pi
            end
        end
        if key == "webhooks"
            for (k,v) in value
                pi = PathItem()
                parse!(pi, v)
                api.webhooks[k] = pi
            end
        end
        if key == "components" parse!(api.components, value) end
        if key == "security" parse!(api.security, value) end
        if key == "tags" parse!(api.tags, value) end
        if key == "externalDocs" parse!(api.externalDocs, value) end
        if key == "x-tagGroups" parse!(api.tagGroups, value) end
    end

    api
end

File: src/parser/openapi.cpp
//
// Created by Rakesh on 19/09/2025.
//

#include "parser.hpp"
#include "model/openapi.hpp"

template <>
void spt::parser::parse( model::Components& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "schemas" )
    {
      for ( const auto& cn : child.children() )
      {
        m.schemas.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::Schema>( cn ) );
      }
    }

    if ( child.key() == "responses" )
    {
      for ( const auto& cn : child.children() )
      {
        m.responses.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::Response>( cn ) );
      }
    }

    if ( child.key() == "parameters" )
    {
      for ( const auto& cn : child.children() )
      {
        m.parameters.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::Parameter>( cn ) );
      }
    }

    if ( child.key() == "examples" )
    {
      for ( const auto& cn : child.children() )
      {
        m.examples.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::Example>( cn ) );
      }
    }

    if ( child.key() == "requestBodies" )
    {
      for ( const auto& cn : child.children() )
      {
        m.requestBodies.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::RequestBody>( cn ) );
      }
    }

    if ( child.key() == "headers" )
    {
      for ( const auto& cn : child.children() )
      {
        m.headers.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::Header>( cn ) );
      }
    }

    if ( child.key() == "securitySchemes" )
    {
      for ( const auto& cn : child.children() )
      {
        m.securitySchemes.try_emplace( std::string{ std::string_view{ cn.key() } }, parse<model::SecurityScheme>( cn ) );
      }
    }
  }
}

template <>
void spt::parser::parse( model::OpenAPI& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "openapi" ) child >> m.openapi;
    if ( child.key() == "jsonSchemaDialect" ) child >> m.jsonSchemaDialect;
    if ( child.key() == "info" ) m.info = parse<model::Info>( child );
    if ( child.key() == "externalDocs" ) m.externalDocs = parse<model::ExternalDocumentation>( child );
    if ( child.key() == "components" ) m.components = parse<model::Components>( child );

    if ( child.key() == "paths" )
    {
      for ( const auto& p : child.children() )
      {
        m.paths.try_emplace( std::string{ std::string_view{ p.key() } }, parse<model::PathItem>( p ) );
      }
    }

    if ( child.key() == "webhooks" )
    {
      for ( const auto& p : child.children() )
      {
        m.webhooks.try_emplace( std::string{ std::string_view{ p.key() } }, parse<model::PathItem>( p ) );
      }
    }

    if ( child.key() == "security" )
    {
      m.security.reserve( 8 );
      for ( const auto& sec : child.children() )
      {
        m.security.emplace_back( parse<model::SecurityRequirement>( sec ) );
      }
    }

    if ( child.key() == "servers" )
    {
      m.servers.reserve( 8 );
      for ( const auto& srv : child.children() )
      {
        m.servers.emplace_back( parse<model::Server>( srv ) );
      }
    }

    if ( child.key() == "tags" )
    {
      m.tags.reserve( 8 );
      for ( const auto& tag : child.children() )
      {
        m.tags.emplace_back( parse<model::Tag>( tag ) );
      }
    }

    if ( child.key() == "x-tagGroups" )
    {
      m.tagGroups.reserve( 8 );
      for ( const auto& tag : child.children() )
      {
        m.tagGroups.emplace_back( parse<model::TagGroup>( tag ) );
      }
    }
  }
}
